lib/events/event.ex

```diff
@@ -0,0 +1,74 @@
+defmodule ElixirTools.Events.Event do
+  @moduledoc """
+  A structure for working with events
+  """
+  alias __MODULE__
+
+  @type t :: %Event{
+          name: String.t(),
+          version: String.t(),
+          payload: map
+        }
+
+  @typep return :: :ok | {:error, reason :: String.t()}
+
+  @enforce_keys ~w(name)a
+  defstruct [:name, payload: %{}, version: "1.0.0"]
+
+  @spec publish(t, module | nil) :: return
+  def publish(event, adapter \\ nil) do
+    adapter = adapter || Application.get_env(:pagantis_elixir_tools, ElixirTools.Events)[:adapter]
+
+    with :ok <- validate(event) do
+      try do
+        adapter.publish(event)
+      rescue
+        e -> {:error, e}
+      end
+    end
+  end
+
+  @spec validate(t) :: return
+  def validate(event) do
+    with :ok <- validate_name(event.name),
+         :ok <- validate_payload(event.payload),
+         :ok <- validate_version(event.version) do
+      :ok
+    end
+  end
+
+  @spec validate_name(any) :: return
+  defp validate_name(name) do
+    cond do
+      !is_binary(name) ->
```

Comment:
What's the bang doing `!is_binary(name)` exactly?

Reply:
negating the return of the `is_binary`, instead of `true` it will return `false`